### ISSUE TITLE
Cascading asset dependencies

### ISSUE THREAD

**Issue by Frooxius** [Yellow-Dog-Man/Resonite-Issues#2737]
### Is your feature request related to a problem? Please describe.

Resonite's asset system tracks the actual usage of assets in the worlds and determines if they should be loaded or not. However currently this only considers the most immediate (direct) references. 

This has a problem where if one asset references another asset (e.g. a material references a texture), that texture will be considered to be "in use", even if the material itself isn't.

In other words, the "in use" doesn't propagate through the asset dependencies. This causes a number of assets to be loaded when they're not necessary.

### Describe the solution you'd like

Implement system that will cascade the asset dependencies - assets will inform other assets they depend on that they are needed only if they are themselves needed.

Additionally, a number of other components - e.g. MeshRenderer, will notify the assets (mesh + materials) that they are not needed under certain conditions - e.g. when it is inactive (under disabled hierarchy), with mechanism to disable this behavior (to allow pre-loading).

### Describe alternatives you've considered

N/A

### Additional Context

This should provide a substantial loading and memory performance boost for worlds and avatars, particularly when they have large parts disabled.

E.g. a lot of users have systems on their avatar that are visible only to the user or only visible when activated.

E.g. private UI's and mechanisms are currently still loaded on other user's ends, even though they can never see them, which contributes to the avatar "load lag".

Similarly if the avatar has different textures, clothing systems and so on that can be swapped, all of them are currently loaded initially, even if they're never needed.

**Comment by TisFoolish** [comment]
> Similarly if the avatar has different textures, clothing systems and so on that can be swapped, all of them are currently loaded initially, even if they're never needed.

To be clear, we'll still download the asset data, it'll just reside in cache until the related slot is enabled?

**Comment by Frooxius** [comment]
No, it won't download at all by default. Assets are only downloaded when they're actually to be loaded.

We will probably add some more mechanisms to preload things too, so there's more control over stuff, but that's a bit separate issue.

**Comment by Azavit** [comment]
I'm curious with how this would work with an asset multiplexer with this change.
Will all the assets count as "loaded" even when it's only using the one who's index is selected?
or will this be smart and not actually load these. 
If not loaded will everything set up in the multiplexer be preloaded? as it's likely to swap quickly between them.

I'm assuming it'll be "smart" about these but I wanted to ask.

**Comment by Frooxius** [comment]
Assets will be requested to be loaded when something that actually uses them is referencing them (and indicating they need them).

So AssetMultiplexer wouldn't cause them to be loaded - unless it mulitplexes the asset to a field on component that will ask that asset to be loaded.

**Comment by Frooxius** [comment]
I'm considering prioritizing this after PhotonDust, before I start work on the audio system.

This is another relatively standalone optimization that I believe would improve overall performance experience significantly, while also not being as complex to implement.

The ways this could help is:
- Loading performance - especially when culling systems are employed - avatars don't need to load on initial join until you get close to them. This will reduce the initial loading time & spike and better structure it
- Hidden systems on avatars - lots of people have stuff they only see locally, but the assets will still load for everyone. This would prevent that from happening, reducing the loading times, but also memory usage by not loading stuff that's not needed for the user
- Worlds with culling systems - they will only need to load the parts that are immediately visible to the user and then load more incrementally

**Comment by Zyzyl** [comment]
This sounds like a great change, but there are some contexts where a change in behaviour might cause issues. 

I'm thinking particularly about the Metamovie, which has the majority of the world under inactive hierarchies before the performance starts. My worry is that, if most of the world's assets aren't downloaded and cached at the start (due to not being in use), participants might see meshes and materials popping in (or audio/video not playing) each time they transition to a different zone. Having options for controlling when assets are downloaded would be really important for us if this were implemented.

**Comment by Frooxius** [comment]
@Zyzyl I'm aware of this causing these kinds of issues potentially, which is why the original posts mentions "with mechanism to disable this behavior (to allow pre-loading)."

It's already planned to have mechanisms to control this behavior, which will resolve these kinds of problems and give creators more control.

**Comment by 5H4D0W-X** [comment]
I would also like separate controls for pre-loading depending on whether the asset is already downlaoded (i.e. found in cache) or not. It would be nice if assets could always download immediately, but would then only be cached without ever loading into working memory (or rather getting cleared from memory after the download finishes, but that's beside the point). This should speed up apparent load times, for example in worlds where rooms are culled based on doors being open, since in that case it's critical that the rooms don't take a while to download after being opened.

**Comment by Frooxius** [comment]
My plan is to have mechanism that lets you indicate that assets should be pre-downloaded, but not actually loaded, but this is also a bit more complicated for things, since it requires resolving variants, which sometimes happens as part of loading.

However this won't likely be default, because it would defeat purpose of this - you'd still be downloading assets you don't really need at the time and won't really need to load.

**Comment by 5H4D0W-X** [comment]
Will there also be options to unload (clear from RAM) assets after they're not in use for a while (or even immediately)? For big worlds, especially with linear progression game worlds where the player cannot travel backwards, unloading past areas would allow for much bigger environments without overly high RAM usage. I'm guessing garbage collection would already kinda do that but that only happens rarely and when it does it often lags the game a fair bit.

**Comment by Frooxius** [comment]
Not explicitly for this. Assets are unloaded automatically when nothing no longer uses them.

If you deactivate stuff that uses them, the system will be able to unload them at that point.
